distribution_metadata.c

```diff
@@ -755,65 +675,61 @@ InsertShardPlacementRow(uint64 shardPlacementId, uint64 shardId,
-	Relation heapRelation = NULL;
-	Relation indexRelation = NULL;
-	IndexScanDesc indexScanDesc = NULL;
-	const int scanKeyCount = 1;
-	ScanKeyData scanKey[scanKeyCount];
-	HeapTuple heapTuple = NULL;
-
-	heapRangeVar = makeRangeVar(METADATA_SCHEMA_NAME, SHARD_PLACEMENT_TABLE_NAME, -1);
-	indexRangeVar = makeRangeVar(METADATA_SCHEMA_NAME,
-								 SHARD_PLACEMENT_PKEY_INDEX_NAME, -1);
-
-	heapRelation = relation_openrv(heapRangeVar, RowExclusiveLock);
-	indexRelation = relation_openrv(indexRangeVar, AccessShareLock);
-
-	ScanKeyInit(&scanKey[0], 1, BTEqualStrategyNumber, F_INT8EQ,
-				Int64GetDatum(shardPlacementId));
-
-	indexScanDesc = index_beginscan(heapRelation, indexRelation, SnapshotSelf,
-									scanKeyCount, 0);
-	index_rescan(indexScanDesc, scanKey, scanKeyCount, NULL, 0);
-
-	heapTuple = index_getnext(indexScanDesc, ForwardScanDirection);
-	if (HeapTupleIsValid(heapTuple))
-	{
-		simple_heap_delete(heapRelation, &heapTuple->t_self);
-	}
-	else
+	Oid argTypes[] = { INT8OID };
+	Datum argValues[] = { Int64GetDatum(shardPlacementId) };
+	const int argCount = sizeof(argValues) / sizeof(argValues[0]);
+	int spiStatus = 0;
+
+	SPI_connect();
+
+	spiStatus = SPI_execute_with_args("DELETE FROM "
+									  "pgs_distribution_metadata.shard_placement "
+									  "WHERE id = $1", argCount, argTypes, argValues,
+									  NULL, false, 0);
+	Assert(spiStatus == SPI_OK_DELETE);
+
+	if (SPI_processed != 1)
 	{
 		ereport(ERROR, (errcode(ERRCODE_UNDEFINED_OBJECT),
 						errmsg("shard placement with ID " INT64_FORMAT " does not exist",
 							   shardPlacementId)));
 	}
 
-	index_endscan(indexScanDesc);
-	index_close(indexRelation, AccessShareLock);
-	relation_close(heapRelation, RowExclusiveLock);
+	SPI_finish();
 }
 
 
 /*
- * NextSequenceId allocates and returns a new unique id generated from the given
- * sequence name.
+ * UpdateShardPlacementRowState sets the shard state of the row identified by
+ * the provided shard placement identifier, erroring out if it cannot find such
+ * a row.
  */
-uint64
-NextSequenceId(char *sequenceName)
+void
+UpdateShardPlacementRowState(int64 shardPlacementId, ShardState newState)
 {
-	RangeVar *sequenceRangeVar = makeRangeVar(METADATA_SCHEMA_NAME,
-											  sequenceName, -1);
-	bool failOk = false;
-	Oid sequenceRelationId = RangeVarGetRelid(sequenceRangeVar, NoLock, failOk);
-	Datum sequenceRelationIdDatum = ObjectIdGetDatum(sequenceRelationId);
-
-	/* generate new and unique id from sequence */
-	Datum sequenceIdDatum = DirectFunctionCall1(nextval_oid, sequenceRelationIdDatum);
-	uint64 nextSequenceId = (uint64) DatumGetInt64(sequenceIdDatum);
+	Oid argTypes[] = { INT8OID, INT4OID };
+	Datum argValues[] = {
+		Int64GetDatum(shardPlacementId),
+		Int32GetDatum((int32) newState)
+	};
+	const int argCount = sizeof(argValues) / sizeof(argValues[0]);
+	int spiStatus = 0;
+
+	SPI_connect();
+
+	spiStatus = SPI_execute_with_args("UPDATE pgs_distribution_metadata.shard_placement "
```

Comment:
This is somewhat unrelated to this change, but how do we manage consistency of locking between Citus and pg_shard? If we are now accessing CitusDB's metadata, Citus acquires locks for certain metadata operations, but pg_shard doesn't seem to? (Or is update getting a lock on the relation for us?)